pkg/services/sqlstore/stats.go

```diff
@@ -192,23 +182,108 @@ func GetSystemUserCountStats(ctx context.Context, query *models.GetSystemUserCou
 	})
 }
 
-func GetActiveUserStats(query *models.GetActiveUserStatsQuery) error {
-	activeUserDeadlineDate := time.Now().Add(-activeUserTimeLimit)
-	sb := &SqlBuilder{}
+func GetUserStats(query *models.GetUserStatsQuery) error {
+	err := updateUserRoleCountsIfNecessary(query.MustUpdate)
+	if err != nil {
+		return err
+	}
 
-	sb.Write(`SELECT `)
-	sb.Write(`(SELECT COUNT(*) FROM `+dialect.Quote("user")+` WHERE last_seen_at > ?) AS active_users,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Viewer", "viewers", true)+`,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Editor", "editors", true)+`,`, activeUserDeadlineDate)
-	sb.Write(roleCounterSQL("Admin", "admins", true)+``, activeUserDeadlineDate)
+	if query.Active {
+		query.Result = userStatsCache.active
+	} else {
+		query.Result = userStatsCache.total
+	}
 
-	var stats models.ActiveUserStats
-	_, err := x.SQL(sb.GetSqlString(), sb.params...).Get(&stats)
+	return nil
+}
+
+func updateUserRoleCountsIfNecessary(forced bool) error {
+	memoizationPeriod := time.Now().Add(-userStatsCacheLimetime)
+	if forced || userStatsCache.memoized.Before(memoizationPeriod) {
+		err := updateUserRoleCounts()
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+type memoUserStats struct {
+	active models.UserStats
+	total  models.UserStats
+
+	memoized time.Time
+}
+
+var (
+	userStatsCache         = memoUserStats{}
+	userStatsCacheLimetime = 5 * time.Minute
+)
+
+func updateUserRoleCounts() error {
+	query := `
+SELECT role AS bitrole, active, COUNT(role) AS count FROM
+    (SELECT active, SUM(role) AS role
+     FROM (SELECT
+               u.id,
+               CASE org_user.role
+                   WHEN 'Admin' THEN 4
+                   WHEN 'Editor' THEN 2
+                   ELSE 1
+                   END AS 'role',
+               CASE
+                WHEN u.last_seen_at>? THEN ` + dialect.BooleanStr(true) + `
+                ELSE ` + dialect.BooleanStr(false) + `
+               END AS active
+           FROM user AS u LEFT JOIN org_user ON org_user.user_id = u.id
+           GROUP BY u.id, u.last_seen_at, org_user.role)
+     GROUP BY active, id)
+GROUP BY active, role;`
```

Comment:
Smart! I think with a few adjustments works against all the databases (sqlite, postgres, mysql)
```suggestion
	userTable := "user"
	if x.Dialect().DriverName() == migrator.POSTGRES {
		userTable = `"user"`
	}
	query := fmt.Sprintf(`
SELECT role AS bitrole, active, COUNT(role) AS count FROM
    (SELECT active, SUM(role) AS role
     FROM (SELECT
               u.id,
               CASE org_user.role
                   WHEN 'Admin' THEN 4
                   WHEN 'Editor' THEN 2
                   ELSE 1
                   END AS role,
               CASE
                WHEN u.last_seen_at>? THEN `+dialect.BooleanStr(true)+`
                ELSE `+dialect.BooleanStr(false)+`
               END AS active
           FROM %s AS u LEFT JOIN org_user ON org_user.user_id = u.id
           GROUP BY u.id, u.last_seen_at, org_user.role) AS t2
     GROUP BY active, id) AS t1
GROUP BY active, role;`, userTable)
```